### ISSUE TITLE
Clustergram not reordering heat-map rows

### ISSUE THREAD

**Issue by plpla** [plotly/dash-bio#432]
**Describe the bug**
Clustergram is not using the order provided by the clustering method and dendrogram to display the heat-map.

**To Reproduce**
```
import numpy as np
import dash_bio
from plotly.offline import iplot, init_notebook_mode

init_notebook_mode()
data = np.array([[1,1,1,1], 
        [3,3,3,3],
        [1,1,1,1],
        [3,3,3,3],
        [1,1,1,1],
        [3,3,3,3]])

iplot(dash_bio.Clustergram(
    data=np.array(data),
    center_values=False
))
```

**Expected behavior**
Heat-map rows reordering based on the clustering result.

**Screenshots**
If applicable, add screenshots to help explain the issue.
![image](https://user-images.githubusercontent.com/1944849/67338652-9414c780-f4f7-11e9-96da-0e03f5e06e96.png)

**Python version: [e.g., 3.7.2]**
3.7.4
**Python environment (all installed packages in your current environment):**
 - 

**Additional context**

**Comment by mkcor** [comment]
Hi @plpla,

Thank you very much for reporting.

Indeed, this is a bug in the [`read_soft()`](https://github.com/plotly/dash-bio-utils/blob/7b9b43a84fe6f091a3eff9a8b6637793dc6a53dc/dash_bio_utils/gene_expression_reader.py#L14) function from `dash-bio-utils`, more specifically when setting argument `return_filtered_data=True`. This in turn calls internal function [`_get_selected_data`](https://github.com/plotly/dash-bio-utils/blob/7b9b43a84fe6f091a3eff9a8b6637793dc6a53dc/dash_bio_utils/gene_expression_reader.py#L189), which gets the dataframe's values (https://github.com/plotly/dash-bio-utils/blob/7b9b43a84fe6f091a3eff9a8b6637793dc6a53dc/dash_bio_utils/gene_expression_reader.py#L209), discarding the dataframe's labeled, ordered structure -- preserving it is the entire point of using dataframes though!!

I noticed this bug when writing tests for the `read_soft()` utility function with `return_filtered_data=True`. The order isn't deterministic, so I couldn't actually write a test for it! We have to fix the function first. I'll be on it next week. If you want to give it a try, I'll be happy to review your work and iterate.

Best,
Marianne

**Comment by mkcor** [comment]
Hi @plpla,

Sorry, I got completely mixed up with a bug in the Clustergram demo app! Here, we are talking about the Clustergram *component*. I'm not sure what you mean by "the order provided by the clustering method and dendrogram..."

By default, the `Clustergram()` method clusters along *columns* first:
https://github.com/plotly/dash-bio/blob/74e51737830ac08da3799189ccb41d65009e0ceb/dash_bio/component_factory/_clustergram.py#L72-L75

In your example, `data` is
```py
array([[1, 1, 1, 1],
       [3, 3, 3, 3],
       [1, 1, 1, 1],
       [3, 3, 3, 3],
       [1, 1, 1, 1],
       [3, 3, 3, 3]])
```
so its transpose matrix is
```py
array([[1, 3, 1, 3, 1, 3],
       [1, 3, 1, 3, 1, 3],
       [1, 3, 1, 3, 1, 3],
       [1, 3, 1, 3, 1, 3]])
```
and we can see that all 'observations' (rows) are identical, so there is zero distance. In the computation of the clusters and dendrogram, this distance matrix is comes first: 
https://github.com/plotly/dash-bio/blob/74e51737830ac08da3799189ccb41d65009e0ceb/dash_bio/component_factory/_clustergram.py#L749-L752

I'm not a [dendrogram](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html) expert, but what kind of dendrogram were you expecting in this trivial case?

**Comment by plpla** [comment]
Hi @mkcor ,
I would expect the result to be highly similar to Searborn Clustermap since they are the same type of figure.
Using the data from my very basic example, the rows with identical values (ones or threes) should be next to each other.  See the example below.
```
import seaborn as sns
import numpy as np

data = np.array([[1,1,1,1], 
        [3,3,3,3],
        [1,1,1,1],
        [3,3,3,3],
        [1,1,1,1],
        [3,3,3,3]])

sns.clustermap(data, cmap="mako")
```

![image](https://user-images.githubusercontent.com/1944849/69995045-4db27f80-151d-11ea-96c9-ee302054f3aa.png)


To be clear, the clusering in the Clustergram works as the dendrogram on the left side shows 2 groups. Although the rows in the heat-map are not reordered to represent these clusters. In the end, I would expect to get a block of red rows and a block of green rows.

**Comment by mkcor** [comment]
Hi @plpla,

Thank you very much for the detailed answer. This is definitely a bug on our end. Our `Clustergram` builds on top of [scipy.cluster.hierarchy](https://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html) just like seaborn's [clustermap](https://seaborn.pydata.org/generated/seaborn.clustermap.html) and we should definitely get two clusters when running `dash_bio.Clustergram(data)` or `dash_bio.Clustergram(data, cluster='row')` (in this example).

I'll try and fix it at the beginning of 2020!

/cc @shammamah

**Comment by mkcor** [comment]
I have finally found the bug, yay! Fixed in 611a1b358b66694a047e9ddafbf9ea7f9562eea1. PR coming soon...
